cnxarchive/views/xpath.py

```diff
@@ -0,0 +1,148 @@
+import json
+from urllib import urlencode
+
+import psycopg2
+import psycopg2.extras
+from pyramid import httpexceptions
+from lxml import etree
+from pyramid.settings import asbool
+from pyramid.threadlocal import get_current_registry, get_current_request
+from pyramid.view import view_config
+
+from .. import config
+from ..database import (
+    SQL, get_tree)
+from ..utils import (
+    IdentHashShortId, IdentHashMissingVersion, IdentHashSyntaxError,
+    split_ident_hash, COLLECTION_MIMETYPE, join_ident_hash
+    )
+from .content import HTML_WRAPPER
+from .helpers import get_content_metadata, get_uuid, get_latest_version
+
+
+# #################### #
+#        Helpers       #
+# #################### #
+
+
+def xpath_book(request, uuid, version, return_json=True):
+    """
+    Given a request, book UUID and version:
+
+    returns a JSON object or HTML list of results, each result containing:
+    module_name,
+    module_uuid,
+    xpath_results, an array of strings, each an individual xpath result.
+    """
+
+    xpath_string = request.params.get('q')
+    results = execute_xpath(xpath_string, 'xpath', uuid, version)
+    if return_json:
+        return results
+    else:
+        return xpath_book_html(request, results)
+
+
+def xpath_book_html(request, results):
+    ul = etree.Element('ul')
+    for item in results:
+        li = etree.SubElement(ul, 'li')
+        a = etree.SubElement(li, 'a')
+        a.set('href', '{}?{}'.format(
+            request.route_path('xpath'),
```

Comment:
Can the `<a href="` links point to the destination content rather than an additional query?

From looking at the code it seems that an XPath query on a book will return a list of links to additional XPath queries (`<a href="/xpath?q=...">`) rather than links to the content in the book (`<a href="/contents/{book_uuid}:{page_uuid}#{nearest_element_id}">`)

Reply:
Is this correct? `%(xpath_string)s/ancestor-or-self::*[@id][1]`

Reply:
yes, that should find the nearest id 😄